src/dmd/objc_glue.d

```diff
@@ -38,273 +37,353 @@ import dmd.backend.type;
+    abstract void generateModuleInfo();
 }
 
-// MARK: ObjcSymbols
+private:
 
-__gshared
+extern(C++) final class Unsupported : ObjcGlue
 {
-    bool objc_hasSymbols = false;
-
-    Symbol *objc_smsgSend = null;
-    Symbol *objc_smsgSend_stret = null;
-    Symbol *objc_smsgSend_fpret = null;
-    Symbol *objc_smsgSend_fp2ret = null;
-
-    Symbol *objc_simageInfo = null;
-    Symbol *objc_smoduleInfo = null;
+    override void setupMethodSelector(FuncDeclaration fd, elem** esel)
+    {
+        // noop
+    }
 
-    StringTable *objc_smethVarNameTable = null;
-    StringTable *objc_smethVarRefTable = null;
-}
+    override void setupMethodCall(elem** ec, elem* ehidden, elem* ethis, TypeFunction tf)
+    {
+        assert(0, "Should never be called when Objective-C is not supported");
+    }
 
-static StringTable *initStringTable(StringTable *stringtable)
-{
-    stringtable = new StringTable();
-    stringtable._init();
+    override void setupEp(elem* esel, elem** ep, int reverse)
+    {
+        // noop
+    }
 
-    return stringtable;
+    override void generateModuleInfo()
+    {
+        // noop
+    }
 }
 
-void objc_initSymbols()
+extern(C++) final class Supported : ObjcGlue
 {
-    objc_hasSymbols = false;
+    extern (D) this()
+    {
+        Symbols.initialize();
+    }
 
-    objc_smsgSend = null;
-    objc_smsgSend_stret = null;
-    objc_smsgSend_fpret = null;
-    objc_smsgSend_fp2ret = null;
+    override void setupMethodSelector(FuncDeclaration fd, elem** esel)
+    {
+        if (fd && fd.selector && !*esel)
+        {
+            *esel = el_var(Symbols.getMethVarRef(fd.selector.toString()));
+        }
+    }
 
-    objc_simageInfo = null;
-    objc_smoduleInfo = null;
+    override void setupMethodCall(elem** ec, elem* ehidden, elem* ethis, TypeFunction tf)
+    {
+        // make objc-style "virtual" call using dispatch function
+        assert(ethis);
+        Type tret = tf.next;
+        *ec = el_var(Symbols.getMsgSend(tret, ehidden !is null));
+    }
 
-    // clear tables
-    objc_smethVarNameTable = initStringTable(objc_smethVarNameTable);
-    objc_smethVarRefTable = initStringTable(objc_smethVarRefTable);
+    override void setupEp(elem* esel, elem** ep, int leftToRight)
+    {
+        if (esel)
+        {
+            // using objc-style "virtual" call
+            // add hidden argument (second to 'this') for selector used by dispatch function
+            if (leftToRight)
+                *ep = el_param(esel, *ep);
+            else
+                *ep = el_param(*ep, esel);
+        }
+    }
 
-    // also wipe out segment numbers
-    for (int s = 0; s < ObjcSegment.SEG_MAX; ++s)
-        objc_segList[s] = 0;
+    override void generateModuleInfo()
+    {
+        if (Symbols.hasSymbols)
+            Symbols.getModuleInfo();
+    }
 }
 
-Symbol *objc_getCString(const(char)* str, size_t len, const(char)* symbolName, ObjcSegment segment)
+struct Segments
```

Comment:
Could have an Objc prefix too, but not necessary.

Reply:
That kind of defeats the purpose of the D modules system.

Reply:
As I said, something that is grep-able or simple to do automated translation with is a nice to have. If another module declares the same struct name, it's not so trivial to determine who uses which version.

But meh.